cmd/kritis/signer/server.go

```diff
@@ -0,0 +1,228 @@
+	Failed    EvaluationStatus = "failed"
+)
+
+type SignRequest struct {
+	Image string
+}
+
+type SignResponse struct {
+	Image      string           `json:"image,omitempty"`
+	Status     EvaluationStatus `json:"status,omitempty"`
+	Message    string           `json:"message,omitempty"`
+	Violations []string         `json:"violations,omitempty"`
+}
+
+// write the sign `response` to `w` with status `code`
+func WriteResponse(w http.ResponseWriter, response SignResponse, code int) {
+	body, err := json.Marshal(response)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	glog.Infof("%s\t%s", response.Image, response.Message)
+	w.Header().Set("Content-Type", "application/json; charset=utf-8")
+	w.WriteHeader(code)
+	w.Write(body)
+}
+
+// Google Pub/Sub message
+type PubSubMessage struct {
+	Message struct {
+		Data []byte `json:"data,omitempty"`
+		ID   string `json:"id"`
+	} `json:"message"`
+	Subscription string `json:"subscription"`
+}
+
+// Google Container Analysis occurrence notification
+type ContainerAnalysisOccurrenceEvent struct {
+	Name             string    `json:"name"`
+	Kind             string    `json:"kind"`
+	NotificationTime time.Time `json:"notificationTime"`
+}
+
+// retrieve the container analysis occurrence associated with this notification
+func GetOccurrence(event ContainerAnalysisOccurrenceEvent) (*grafeaspb.Occurrence, error) {
+	request := grafeaspb.GetOccurrenceRequest{Name: event.Name}
+	response, err := grafeas.GetOccurrence(context.Background(), &request)
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve discovery occurrence %s", event.Name)
+	}
+	return response, nil
+}
+
+// implements `/event` handles a container analysis occurrence notification
+func containerAnalysisEvent(w http.ResponseWriter, r *http.Request) {
+	var m PubSubMessage
+	var event ContainerAnalysisOccurrenceEvent
+	err := json.NewDecoder(r.Body).Decode(&m)
+	if err != nil {
+		http.Error(w, "could not unmarshal the pubsub message", http.StatusBadRequest)
+		return
+	}
+
+	err = json.Unmarshal(m.Message.Data, &event)
+	if err != nil {
+		http.Error(w, "could not unmarshal container analysis occurrence event from pubsub message", http.StatusBadRequest)
+		return
+	}
+
+	if event.Kind != "DISCOVERY" {
+		http.Error(w, "ignoring non-DISCOVERY occurrence", http.StatusOK)
+		return
+	}
+
+	occurrence, err := GetOccurrence(event)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	status := occurrence.GetDiscovered().GetDiscovered().GetAnalysisStatus()
+	if status != discovery.Discovered_FINISHED_SUCCESS {
+		http.Error(w, fmt.Sprintf("ignoring DISCOVERY occurrence in status %s", string(status)), http.StatusOK)
+		return
+	}
+	image := occurrence.Resource.Uri
+	if strings.HasPrefix(image, "https://") {
+		image = image[8:]
+	}
+	request := SignRequest{Image: image}
+	doCheckAndSign(w, request, http.StatusOK)
```

Comment:
I think this will swallow any errors.  Technically it'll return them back to containeranalysis, but I think they won't even be logged anywhere.  Would it be better to at least log them here?

I'm aware this wouldn't help all that much with debugging, because the user would still have to sift through the server logs; but maybe it's better than nothing.

Reply:
I am not quite sure what you are referring to here. When the image does not start with `https://` it will be passed in as is. the DoCheck() function will validate whether it is a proper image reference with digest and return an error if it is not.